gateway/src/apicast/balancer.lua

```diff
@@ -1,18 +1,53 @@
 local round_robin = require 'resty.balancer.round_robin'
+local resty_url = require 'resty.url'
+local empty = {}
 
 local _M = { default_balancer = round_robin.new() }
 
+local function get_default_port(upstream_url)
+  local url = resty_url.split(upstream_url) or empty
+  local scheme = url[1] or 'http'
+  return resty_url.default_port(scheme)
+end
+
+local function exit_service_unavailable()
+  ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
+  ngx.exit(ngx.status)
+end
+
 function _M.call(_, _, balancer)
   balancer = balancer or _M.default_balancer
   local host = ngx.var.proxy_host -- NYI: return to lower frame
   local peers = balancer:peers(ngx.ctx[host])
 
-  local peer, err = balancer:set_peer(peers)
+  local peer, err = balancer:select_peer(peers)
+
+  local address, port
 
   if not peer then
-    ngx.status = ngx.HTTP_SERVICE_UNAVAILABLE
-    ngx.log(ngx.ERR, "failed to set current backend peer: ", err)
-    ngx.exit(ngx.status)
+    ngx.log(ngx.ERR, 'could not select peer: ', err)
+    return exit_service_unavailable()
+  end
+
+  address, port = peer[1], peer[2]
+
+  if not address then
+    ngx.log(ngx.ERR, 'peer missing address')
+    return exit_service_unavailable()
+  end
+
+  if not port then
+    port = get_default_port(ngx.var.proxy_pass)
```

Comment:
I wonder if this can return `nil` because of the limitations of resty_url.default_port that I pointed out in another comment.
Not sure if that would be a problem later on.

Reply:
I guess it can, but not sure what we should do in this case.
Defaulting to some specific port doesn't sound too good either. I guess here we can accept that apicast will return a 5xx error.